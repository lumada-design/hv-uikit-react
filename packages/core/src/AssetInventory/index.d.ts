export { default as ListView } from "./ListView";
export * from "./ListView";

export { default as CardView } from "./CardView";
export * from "./CardView";

export { default } from "./AssetInventory";
export * from "./AssetInventory";
