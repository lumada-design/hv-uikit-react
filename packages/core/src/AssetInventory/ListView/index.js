import ListViewContext, {
  ListViewContextProvider,
  ListViewContextConsumer,
} from "./ListViewContext/ListViewContext";

export { default } from "./ListView";
export { default as HvListViewRow } from "./ListViewRow";
export { default as HvListViewCell } from "./ListViewCell";
export { ListViewContext as HvListViewContext };
export { ListViewContextProvider as HvListViewContextProvider };
export { ListViewContextConsumer as HvListViewContextConsumer };
