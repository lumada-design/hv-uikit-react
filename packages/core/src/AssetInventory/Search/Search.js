/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import PropTypes from "prop-types";
import get from "lodash/get";
import SearchBox from "../../SearchBox";

/**
 * Filter the metadata that has searchable definitions.
 *
 * @param metadata
 * @returns {[]}
 */
const searchOperationSetup = metadata => {
  const searchableCriteria = [];

  metadata.forEach(element => {
    if (element.searchable) {
      searchableCriteria.push(element);
    }
  });

  return searchableCriteria;
};

/**
 * Filter the search string, based in the criteria, using the value.
 *
 * @param value
 * @param criteria
 * @param searchValue
 * @returns {boolean|*}
 */
const filter = (value, criteria, searchValue) => {
  const evalValue = get(value, criteria.accessor);

  if (criteria.searchFunction)
    return criteria.searchFunction(evalValue, searchValue);

  switch (criteria.cellType.toUpperCase()) {
    case "NUMERIC":
      return evalValue.toString().includes(searchValue.toString());
    case "ALPHA-NUMERIC":
      return evalValue.toUpperCase().includes(searchValue.toUpperCase());
    case "DATE":
      return new Date(evalValue).getTime() === new Date(searchValue).getTime();
    default:
      return value === searchValue;
  }
};

/**
 * For each criteria the value is evaluated, returning a list of filtered results.
 *
 * @param searchValue
 * @param values
 * @param searchableCriteria
 * @returns {[]}
 */
const searchOperation = (searchValue, values, searchableCriteria) => {
  const filteredValues = [];

  searchableCriteria.forEach(criteria => {
    values.forEach(value => {
      if (
        filter(value, criteria, searchValue) &&
        !filteredValues.includes(value)
      )
        filteredValues.push(value);
    });
  });

  return filteredValues;
};

/**
 * Search component.
 *
 * @param id
 * @param labels
 * @param onFilter
 * @param values
 * @param metadata
 * @returns {*}
 * @constructor
 */
const Search = ({ id, searchBoxLabels, onFilter, onSearch, values, metadata }) => {
  const searchableCriteria = searchOperationSetup(metadata);

  const handler = value => {
    const filteredResults = searchOperation(value, values, searchableCriteria);
    onFilter(filteredResults);
    return value;
  };

  return (
    <SearchBox
      id={`search_${id}`}
      labels={searchBoxLabels}
      onChange={onSearch || handler}
    />
  );
};

Search.propTypes = {
  /**
   * Id to be applied to the root node.
   */
  id: PropTypes.string,
  /**
   * Data passed to the component.
   */
  values: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string
    })
  ).isRequired,
  /**
   * Labels.
   */
  searchBoxLabels: PropTypes.shape({
    sortBy: PropTypes.string
  }).isRequired,
  /**
   * Metadata associated with the values.
   */
  metadata: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      accessor: PropTypes.string,
      cellType: PropTypes.oneOf(["alpha-numeric", "numeric", "date", "node"]),
      sortable: PropTypes.bool,
      sortFunction: PropTypes.func,
      searchable: PropTypes.bool,
      searchFunction: PropTypes.func
    })
  ).isRequired,
  /**
   * Callback function returning the filtered data.
   */
  onFilter: PropTypes.func.isRequired,
  /**
   * onSearch callback.
   */
  onSearch: PropTypes.func
};

Search.defaultProps = {
  id: undefined,
  onSearch: null
};

export default Search;
export { searchOperation, filter, searchOperationSetup };
