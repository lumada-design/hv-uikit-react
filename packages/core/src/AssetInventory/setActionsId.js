import { setId } from "../utils";

/**
 * Sets individual ids for each action, using the action id and the data id.
 *
 * @param actions
 * @param id
 * @returns {*}
 */
const setActionsId = (actions, id) => {
  return actions?.map((action) => ({ ...action, id: setId(action.id, id) }));
};

export default setActionsId;
