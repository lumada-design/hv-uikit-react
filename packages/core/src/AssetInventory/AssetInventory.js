import React from "react";
import PropTypes from "prop-types";
import find from "lodash/find";
import isNil from "lodash/isNil";
import isEqual from "lodash/isEqual";
import sort from "lodash/sortBy";
import isEmpty from "lodash/isEmpty";
import { withStyles } from "@material-ui/core";
import clsx from "clsx";
import MultiButton from "./Multibutton/Multibutton";
import withLabels from "../withLabels";
import Search from "./Search/Search";
import Sort from "./Sort/Sort";
import Grid from "../Grid";
import Pagination from "../Pagination";
import styles from "./styles";
import { setId } from "../utils";

// TODO: review event args

const DEFAULT_LABELS = {
  sortBy: "Sort by",
  inputLabel: "",
  placeholder: "Search",
};

/**
 * An Asset Inventory allows to switch between views. The Sort and Filter are defined using the metadata configuration, while the remaining configuration can be ser in the AssetInventory or in the individual views.
 */
class AssetInventory extends React.Component {
  static areArraysEquals(a1, a2) {
    return isEqual(sort(a1), sort(a2));
  }

  constructor(props) {
    super(props);
    const {
      values,
      pageSizeOptions,
      page,
      pageSize,
      selectedValues,
      selectedView,
      children,
      searchString,
      sortOptionId,
    } = this.props;

    const innerPageSize = pageSize || pageSizeOptions[0];
    const viewValues = this.getPaginationData(values, innerPageSize, page);
    const selectedViewId =
      selectedView || (Array.isArray(children) ? children[0].props.id : children.props.id);
    this.state = {
      selectedView: selectedViewId,
      originalSelectedView: selectedView,
      pageSize: innerPageSize,
      originalPageSize: innerPageSize,
      page,
      originalPage: page,
      // original values to compare in the getDerivedStateFromProps
      originalValues: values.slice(),
      // Data for manipulation (search and sorts)
      values: values.slice(),
      // Data shown in the "window"
      viewValues,
      // original selectedValues
      originalSelectedValues: selectedValues.slice(),
      // Values already selected
      selectedValues: selectedValues.slice(),
      // search
      searchString,
      originalSearchString: searchString,
      // sort
      selectedSort: sortOptionId,
      originalSelectedSort: sortOptionId,
    };
  }

  static getDerivedStateFromProps(props, state) {
    let result = {};

    if (
      !AssetInventory.areArraysEquals(props.values, state.originalValues) ||
      !AssetInventory.areArraysEquals(props.selectedValues, state.originalSelectedValues)
    ) {
      result = {
        originalValues: props.values,
        values: props.values,
        viewValues: props.values,
        page: props.page,
        selectedValues: !AssetInventory.areArraysEquals(
          props.selectedValues,
          state.originalSelectedValues
        )
          ? props.selectedValues
          : state.selectedValues,
        pageSize: props.pageSize,
      };
    }

    if (props.selectedView !== state.originalSelectedView) {
      result = {
        ...result,
        selectedView:
          props.selectedView !== state.originalSelectedView
            ? props.selectedView
            : state.selectedView,
        originalSelectedView:
          props.selectedView !== state.originalSelectedView
            ? props.selectedView
            : state.originalSelectedView,
      };
    }

    if (props.sortOptionId !== state.originalSelectedSort) {
      result = {
        ...result,
        originalSelectedSort:
          props.sortOptionId !== state.originalSelectedSort
            ? props.sortOptionId
            : state.originalSelectedSort,
        selectedSort:
          props.sortOptionId !== state.originalSelectedSort
            ? props.sortOptionId
            : state.sortOptionId,
      };
    }
    if (props.searchString !== state.originalSearchString) {
      result = {
        ...result,
        originalSearchString:
          props.searchString !== state.originalSearchString
            ? props.searchString
            : state.originalSearchString,
        searchString:
          props.searchString !== state.originalSearchString
            ? props.searchString
            : state.searchString,
      };
    }

    if (props.pageSize !== state.originalPageSize) {
      result = {
        ...result,
        originalPageSize: props.pageSize,
        pageSize: props.pageSize,
      };
    }

    if (props.page !== state.originalPage) {
      result = {
        ...result,
        originalPage: props.page,
        page: props.page,
      };
    }

    return isEmpty(result) ? null : result;
  }

  getPaginationData = (values, pageSize, page) => {
    const { hasPagination, paginationServerSide } = this.props;
    return hasPagination && !paginationServerSide
      ? values.slice(pageSize * page, pageSize * (page + 1))
      : values;
  };

  changeView = (event, id) => {
    const selectedId = id[0];

    this.setState({
      selectedView: selectedId,
    });
  };

  setViewValues = (returnedViewValues) => {
    const { pageSize, page } = this.state;
    this.setState({
      viewValues: this.getPaginationData(returnedViewValues, pageSize, page),
    });
  };

  setSearchResults = (results, value) => {
    this.setState({ values: results, page: 0, searchString: value });
    this.setViewValues(results);
  };

  /**
   * Show the search component.
   *
   * @returns {*}
   */
  renderSearch = () => {
    const { id, values, classes, labels, configuration, onSearch, searchProps } = this.props;
    const { searchString } = this.state;
    const { inputLabel, placeholder } = labels;
    return (
      <div className={classes.searchBoxContainer}>
        <Search
          id={id}
          searchString={searchString}
          values={values}
          metadata={configuration.metadata}
          onFilter={this.setSearchResults}
          onSearch={onSearch}
          labels={{ inputLabel, placeholder }}
          {...searchProps}
        />
      </div>
    );
  };

  onSort = (sortFunc, id) => {
    const { values } = this.state;
    values.sort(sortFunc);
    this.setViewValues(values);
    this.setState({ selectedSort: id });
  };

  renderSort = () => {
    const { id, labels, configuration, classes, onSortChange, disablePortal, sortProps } =
      this.props;
    const { selectedSort } = this.state;
    const dropDownLabel = {
      title: labels.sortBy,
    };

    return (
      <div className={classes.sortContainer}>
        <Sort
          id={id}
          labels={dropDownLabel}
          metadata={configuration.metadata}
          selected={selectedSort}
          onSelection={this.onSort}
          onSortChange={onSortChange}
          disablePortal={disablePortal}
          {...sortProps}
        />
      </div>
    );
  };

  paginationOnPageChange = (page) => {
    const { values, pageSize } = this.state;

    const pageData = values.slice(pageSize * page, pageSize * (page + 1));

    this.setState({
      page,
      viewValues: pageData,
    });
  };

  paginationOnPageSizeChange = (newPageSize) => {
    const { values, page } = this.state;

    const pageData = values.slice(newPageSize * page, newPageSize * (page + 1));

    this.setState({
      pageSize: newPageSize,
      viewValues: pageData,
    });
  };

  renderPagination = () => {
    const {
      id,
      paginationServerSide,
      paginationProps,
      pages,
      pageSizeOptions,
      classes,
      onPageChange,
      onPageSizeChange,
    } = this.props;

    const { pageSize, page, values } = this.state;

    if (values.length === 0) {
      return undefined;
    }

    const numPages = paginationServerSide ? pages : Math.ceil(values.length / pageSize);

    const onPageChangeInternal = paginationServerSide ? onPageChange : this.paginationOnPageChange;

    const onPageSizeChangeInternal = paginationServerSide
      ? onPageSizeChange
      : this.paginationOnPageSizeChange;

    return (
      <Pagination
        id={setId(id, "pagination")}
        classes={{
          root: classes.pagination,
        }}
        pages={numPages}
        page={page}
        canPrevious={page > 0}
        canNext={page < numPages - 1}
        pageSize={pageSize}
        pageSizeOptions={pageSizeOptions}
        onPageChange={onPageChangeInternal}
        onPageSizeChange={onPageSizeChangeInternal}
        {...paginationProps}
      />
    );
  };

  innerOnSelection = (onSelection) => (event) => {
    const { selectedValues } = this.state;
    const id = event.target.value;

    let list;
    if (event.target.checked) {
      list = [...selectedValues, id];
    } else {
      const index = selectedValues.indexOf(id);
      if (index > -1) {
        list = selectedValues.filter((item) => item !== id);
      }
    }
    this.setState({ selectedValues: list });
    onSelection(event);
  };

  propsFillerManager = (source, target, props) => {
    props.forEach((prop) => this.propsFiller(source, target, prop[0], prop[1]));
  };

  propsFiller = (source, target, propName, value) => {
    if (isNil(source.props[propName])) {
      // eslint-disable-next-line no-param-reassign
      target[propName] = value;
    }
  };

  fillChildProp(child) {
    const {
      onSelection,
      isSelectable,
      actions,
      maxVisibleActions,
      actionsCallback,
      configuration,
    } = this.props;

    let childProps = {};

    if (!isNil(child.props) && !isNil(child.props.viewConfiguration)) {
      childProps = { ...child.props.viewConfiguration };
    }

    if (!isNil(configuration) && !isNil(configuration.viewConfiguration)) {
      childProps = { ...configuration.viewConfiguration };
    }

    this.propsFillerManager(child, childProps, [
      ["onSelection", this.innerOnSelection(onSelection)],
      ["isSelectable", isSelectable],
      ["actions", actions],
      ["maxVisibleActions", maxVisibleActions],
      ["actionsCallback", actionsCallback],
    ]);

    return childProps;
  }

  /**
   * Render the view.
   *
   * @returns {*}
   */
  renderView() {
    const { selectedView, viewValues, selectedValues } = this.state;
    const { children } = this.props;

    const view = Array.isArray(children)
      ? children.find((element) => element.props.id === selectedView)
      : children;

    const childProps = this.fillChildProp(view);

    return React.cloneElement(view, {
      values: viewValues,
      selectedValues,
      viewConfiguration: childProps,
    });
  }

  /**
   * Render the passed filter placeholder.
   *
   * @returns {*}
   */
  renderFilterPlaceholder() {
    const { FilterPlaceholder } = this.props;

    return <FilterPlaceholder onSelection={this.setViewValues} />;
  }

  render() {
    const {
      id,
      classes,
      className,
      FilterPlaceholder,
      children,
      configuration,
      hasPagination,
      onViewChange,
      multibuttonProps,
    } = this.props;

    const { selectedView } = this.state;

    const showButtons = children.length > 1;
    const showSort = find(configuration.metadata, (element) => element.sortable);
    const showSearch = find(configuration.metadata, (element) => element.searchable);
    const showRightControls = showButtons || showSort;
    const views = [];

    const align = !showSearch ? "flex-end" : "space-between";

    React.Children.forEach(children, (child) => {
      const other = multibuttonProps.find((elem) => elem.id === child.props.id);
      views.push({
        icon: child.props.icon,
        selected: child.props.id === selectedView,
        ...other,
        id: setId(child.props.id, "button"),
      });
    });

    const pagination = this.renderPagination();

    return (
      <div id={id} className={clsx(className, classes.root)}>
        {FilterPlaceholder && this.renderFilterPlaceholder()}
        <Grid container spacing={0}>
          <Grid container justify={align} alignItems="flex-end">
            {showSearch && <Grid item>{this.renderSearch()}</Grid>}
            {showRightControls && (
              <Grid item>
                <Grid container alignItems="flex-end" spacing={0}>
                  {showSort && <Grid item>{this.renderSort()}</Grid>}
                  {showButtons && (
                    <Grid item>
                      <div className={classes.multiButtons}>
                        <MultiButton
                          id={id}
                          views={views}
                          changeView={this.changeView}
                          onViewChange={onViewChange}
                        />
                      </div>
                    </Grid>
                  )}
                </Grid>
              </Grid>
            )}
          </Grid>
          <div className={classes.viewContainer}>{this.renderView()}</div>
          {hasPagination && (
            <Grid container>
              <Grid item xs={12} sm={12} md={12} lg={12} xl={12}>
                {pagination}
              </Grid>
            </Grid>
          )}
        </Grid>
      </div>
    );
  }
}

AssetInventory.propTypes = {
  /**
   * Class names to be applied.
   */
  className: PropTypes.string,
  /**
   * Id to be applied to the root node.
   */
  id: PropTypes.string,
  /**
   * A Jss Object used to override or extend the component styles applied.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the root component.
     */
    root: PropTypes.string,
    /**
     * Styles applied to the controls container.
     */
    controlsContainer: PropTypes.string,
    /**
     * Styles applied to the search container.
     */
    search: PropTypes.string,
    /**
     * Styles applied to the right controls container.
     */
    rightControls: PropTypes.string,
    /**
     *  Styles applied to the multiButtons.
     */
    multiButtons: PropTypes.string,
    /**
     * Styles applied to the view container.
     */
    viewContainer: PropTypes.string,
    /**
     * Styles applied to the search box container.
     */
    searchBoxContainer: PropTypes.string,
    /**
     * Styles applied to the sort container.
     */
    sortContainer: PropTypes.string,
    /**
     * Styles applied to the pagination component.
     */
    pagination: PropTypes.string,
  }).isRequired,
  /**
   * Views components.
   */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  /**
   * Data passed to the component.
   */
  values: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
    })
  ).isRequired,
  /**
   * Labels.
   */
  labels: PropTypes.shape({
    /**
     * Sort label.
     */
    sortBy: PropTypes.string,
    /**
     * the label on top of the search box.
     */
    inputLabel: PropTypes.string,
    /**
     * the placeholder value of the search box.
     */
    placeholder: PropTypes.string,
  }),
  /**
   * Contains the metadata for the values and the necessary configuration for the views,
   * check the views for the specific view configuration.
   */
  configuration: PropTypes.shape({
    metadata: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        title: PropTypes.string,
        accessor: PropTypes.string,
        cellType: PropTypes.oneOf(["alpha-numeric", "numeric", "date", "node"]),
        sortable: PropTypes.bool,
        sortFunction: PropTypes.func,
        searchable: PropTypes.bool,
        searchFunction: PropTypes.func,
      })
    ).isRequired,
    viewConfiguration: PropTypes.instanceOf(Object),
  }).isRequired,
  /**
   * Callback evoked in the selection of the card.
   */
  onSelection: PropTypes.func,
  /**
   * List of actions to be passed to the views.
   */
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      label: PropTypes.string,
      icon: PropTypes.func,
      disabled: PropTypes.bool,
    })
  ),
  /**
   *  The callback function ran when an action is triggered, receiving ´action´ as param
   */
  actionsCallback: PropTypes.func,
  /**
   *  The number of maximum visible actions before they're collapsed into a ´DropDownMenu´.
   */
  maxVisibleActions: PropTypes.number,
  /**
   * Extra filters
   */
  FilterPlaceholder: PropTypes.element,
  /**
   * Indicates if the views are selectable.
   */
  isSelectable: PropTypes.bool,
  /**
   * Values selected. The list can be maintain internally or it can be passed (overwriting the internal).
   */
  selectedValues: PropTypes.arrayOf(PropTypes.string),
  /**
   * The selected view id.
   */
  selectedView: PropTypes.string,
  /**
   * Defines if it has pagination.
   */
  hasPagination: PropTypes.bool,
  /**
   * Enable or disable the server side pagination mechanism
   */
  paginationServerSide: PropTypes.bool,
  /**
   * The array of possible page sizes for the dropdown.
   */
  pageSizeOptions: PropTypes.arrayOf(PropTypes.number),
  /**
   * Page size.
   */
  pageSize: PropTypes.number,
  /**
   * The currently selected page (0-indexed).
   */
  page: PropTypes.number,
  /**
   * The number of pages the component has.
   */
  pages: PropTypes.number,
  /**
   * Page change callback.
   */
  onPageChange: PropTypes.func,
  /**
   * Page size change callback.
   */
  onPageSizeChange: PropTypes.func,
  /**
   * Search callback.
   */
  onSearch: PropTypes.func,
  /**
   * Sort callback.
   */
  onSortChange: PropTypes.func,
  /**
   * View change callback.
   */
  onViewChange: PropTypes.func,
  /**
   * Visual indication of the sort applied. The id is given by the metadata.id+Asc or metadata.id+Desc.
   */
  sortOptionId: PropTypes.string,
  /**
   * Visual indicator of the search string used.
   */
  searchString: PropTypes.string,
  /**
   * Disable portal on the Sort dropdown
   */
  disablePortal: PropTypes.bool,
  /**
   * Other props passed to the searchbox.
   */
  searchProps: PropTypes.instanceOf(Object),
  /**
   * Others props passed to the Sort. If you want to control the aria-label
   * use the labels.sortBy, as it is mapped directly to the aria-label.
   */
  sortProps: PropTypes.instanceOf(Object),
  /**
   * Array of others prop passed to the created button. Each element must include the id of the view
   * and other props to pe passed to each button.
   */
  multibuttonProps: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
    })
  ),

  /**
   *  Extra properties passed to the pagination.
   */
  paginationProps: PropTypes.instanceOf(Object),
};

AssetInventory.defaultProps = {
  className: "",
  id: undefined,
  FilterPlaceholder: null,
  onSelection: null,
  actions: null,
  actionsCallback: null,
  maxVisibleActions: 1,
  isSelectable: false,

  selectedValues: [],
  selectedView: null,
  hasPagination: false,
  paginationServerSide: false,
  pageSizeOptions: [5, 10, 20, 25, 50, 100],
  page: 0,
  pages: 0,
  onPageChange: null,
  onPageSizeChange: null,
  pageSize: undefined,
  onSearch: null,
  onSortChange: null,
  onViewChange: () => {},
  sortOptionId: null,
  searchString: undefined,
  disablePortal: false,
  searchProps: undefined,
  sortProps: undefined,
  multibuttonProps: [],
  paginationProps: undefined,
};

export default withStyles(styles, { name: "HvAssetInventory" })(
  withLabels(DEFAULT_LABELS)(AssetInventory)
);
