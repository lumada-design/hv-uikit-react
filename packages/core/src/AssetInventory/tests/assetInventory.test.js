/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { mount, shallow } from "enzyme";
import HvProvider from "../../Provider";
import AssetInventory from "../index";
import Search from "../Search/Search";

describe("Asset Inventory ", () => {
  let wrapper;

  const values = [
    {
      id: "1",
      name: "AA",
      birthday: "1900-01-01",
      number: 20
    },
    {
      id: "2",
      name: "HH",
      birthday: "1950-01-01",
      number: 92
    },
    {
      id: "3",
      name: "ZZ",
      birthday: "1923-01-01",
      number: 1
    }
  ];

  beforeEach(() => {
    wrapper = shallow(
      <HvProvider>
        <AssetInventory values={values} configuration={[]}>
          <div id="id0" />
        </AssetInventory>
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correctly sort", () => {
    const configuration = {
      metadata: [
        {
          id: "id1",
          accessor: "name",
          cellType: "alpha-numeric",
          sortable: true,
          sortableLabelAsc: "Name ascending",
          sortableLabelDesc: "Name descending"
        }
      ]
    };

    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={configuration}>
          <div id="id" />
        </AssetInventory>
      </HvProvider>
    );

    const sort = wrapper.find("Sort");

    expect(sort.length).toBe(1);
  });

  it("shouldn't render sort", () => {
    const configuration = {
      metadata: [
        {
          id: "id1",
          accessor: "name",
          cellType: "alpha-numeric"
        }
      ]
    };

    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={configuration}>
          <div id="id" />
        </AssetInventory>
      </HvProvider>
    );

    const sort = wrapper.find("Sort");

    expect(sort.length).toBe(0);
  });

  it("should render correctly search", () => {
    const configuration = {
      metadata: [
        {
          id: "id1",
          accessor: "name",
          cellType: "alpha-numeric",
          searchable: true
        }
      ]
    };

    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={configuration}>
          <div id="id" />
        </AssetInventory>
      </HvProvider>
    );

    const search = wrapper.find(Search);

    expect(search.length).toBe(1);
  });

  it("shouldn't render correctly search", () => {
    const configuration = {
      metadata: [
        {
          id: "id1",
          accessor: "name",
          cellType: "alpha-numeric"
        }
      ]
    };

    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={configuration}>
          <div id="id" />
        </AssetInventory>
      </HvProvider>
    );

    const search = wrapper.find("Search");

    expect(search.length).toBe(0);
  });

  it("should correctly filter alpha-numeric values", () => {
    const configuration = {
      metadata: [
        {
          id: "id1",
          accessor: "name",
          cellType: "alpha-numeric",
          searchable: true
        }
      ]
    };

    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={configuration}>
          <div id="id" />
        </AssetInventory>
      </HvProvider>
    );

    let instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(3);

    const search = wrapper.find(Search).find("input");
    search.simulate("change", { target: { value: "a" } });

    instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(1);
    expect(instance.state.viewValues[0].name).toBe("AA");
  });

  it("should correctly filter numeric values", () => {
    const configuration = {
      metadata: [
        {
          id: "id1",
          accessor: "number",
          cellType: "numeric",
          searchable: true
        }
      ]
    };

    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={configuration}>
          <div id="id" />
        </AssetInventory>
      </HvProvider>
    );

    let instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(3);

    const search = wrapper.find(Search).find("input");
    search.simulate("change", { target: { value: 2 } });

    instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(2);
  });

  it("should correctly filter custom values", () => {
    const configuration = {
      metadata: [
        {
          id: "id1",
          accessor: "number",
          cellType: "numeric",
          searchable: true,
          searchFunction: (evalValue, searchValue) => evalValue > searchValue
        }
      ]
    };

    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={configuration}>
          <div id="id" />
        </AssetInventory>
      </HvProvider>
    );

    let instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(3);

    let search = wrapper.find(Search).find("input");
    search.simulate("change", { target: { value: 0 } });

    instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(3);

    search = wrapper.find(Search).find("input");
    search.simulate("change", { target: { value: 10 } });

    instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(2);

    search = wrapper.find(Search).find("input");
    search.simulate("change", { target: { value: 50 } });

    instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(1);
  });

  it("should correctly switch views", () => {
    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={[]}>
          <div id="view1">test1</div>
          <div id="view2" />
        </AssetInventory>
      </HvProvider>
    );

    let view = wrapper.findWhere(
      n => n.type() === "div" && n.prop("id") === "view1"
    );

    expect(view.exists()).toBe(true);

    view = wrapper.findWhere(
      n => n.type() === "div" && n.prop("id") === "view2"
    );

    expect(view.exists()).toBe(false);

    wrapper
      .findWhere(n => n.type() === "button" && n.prop("id") === "view2")
      .simulate("click");

    view = wrapper.findWhere(
      n => n.type() === "div" && n.prop("id") === "view2"
    );

    expect(view.exists()).toBe(true);

    view = wrapper.findWhere(
      n => n.type() === "div" && n.prop("id") === "view1"
    );

    expect(view.exists()).toBe(false);
  });

  it("should render pagination", () => {
    wrapper = mount(
      <HvProvider>
        <AssetInventory values={values} configuration={{}} hasPagination>
          <div id="view1">test1</div>
        </AssetInventory>
      </HvProvider>
    );

    const pagination = wrapper.find("Pagination");

    expect(pagination.length).toBe(1);
  });

  it("should present a subset of data using pagination, calculating the page total", () => {
    const val = [
      {
        id: "1",
        name: "AA"
      },
      {
        id: "2",
        name: "BB"
      },
      {
        id: "3",
        name: "CC"
      },
      {
        id: "4",
        name: "DD"
      },
      {
        id: "5",
        name: "EE"
      },
      {
        id: "6",
        name: "FF"
      }
    ];

    wrapper = mount(
      <HvProvider>
        <AssetInventory
          id="hv-pagination"
          values={val}
          configuration={{}}
          hasPagination
          pageSize={2}
        >
          <div id="view1">test1</div>
        </AssetInventory>
      </HvProvider>
    );

    const instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues.length).toBe(2);

    const totalPages = wrapper.findWhere(
      n => n.type() === "span" && n.prop("id") === "hv-pagination-totalPages"
    );

    expect(totalPages.text()).toBe("3");
  });

  it("should change the subset of data using pagination", () => {
    const val = [
      {
        id: "1",
        name: "AA"
      },
      {
        id: "2",
        name: "BB"
      },
      {
        id: "3",
        name: "CC"
      },
      {
        id: "4",
        name: "DD"
      },
      {
        id: "5",
        name: "EE"
      },
      {
        id: "6",
        name: "FF"
      }
    ];

    wrapper = mount(
      <HvProvider>
        <AssetInventory
          id="hv-pagination"
          values={val}
          configuration={{}}
          hasPagination
          pageSize={2}
        >
          <div id="view1">test1</div>
        </AssetInventory>
      </HvProvider>
    );

    const instance = wrapper.find("AssetInventory").instance();

    expect(instance.state.viewValues[0].id).toBe("1");
    expect(instance.state.viewValues[1].id).toBe("2");

    wrapper
      .findWhere(
        n =>
          n.type() === "button" &&
          n.prop("id") === "hv-pagination-nextPage-button"
      )
      .simulate("click");

    expect(instance.state.viewValues.length).toBe(2);

    expect(instance.state.viewValues[0].id).toBe("3");
    expect(instance.state.viewValues[1].id).toBe("4");
  });
});
