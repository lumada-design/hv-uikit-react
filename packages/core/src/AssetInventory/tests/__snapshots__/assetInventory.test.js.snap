// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset Inventory  should render correctly 1`] = `
<JssProvider
  generateClassName={[Function]}
>
  <HvProvider
    changeTheme={[Function]}
    router={null}
    theme={null}
    uiKitTheme="dawn"
  >
    <WithStyles(AssetInventory)
      configuration={Array []}
      values={
        Array [
          Object {
            "birthday": "1900-01-01",
            "id": "1",
            "name": "AA",
            "number": 20,
          },
          Object {
            "birthday": "1950-01-01",
            "id": "2",
            "name": "HH",
            "number": 92,
          },
          Object {
            "birthday": "1923-01-01",
            "id": "3",
            "name": "ZZ",
            "number": 1,
          },
        ]
      }
    >
      <div
        id="id0"
      />
    </WithStyles(AssetInventory)>
  </HvProvider>
</JssProvider>
`;
